### PR TITLE
Unused variable detection

### PR DESCRIPTION
**Pull request by LucasSte**
This pull request implements unused variable detection for the Solang compiler.
It generates warnings for unused state variables, global constants and contract storage.
As a bonus, I implemented the detection of events that have never been emitted.

This PR also implements tests for the new code.

This is the first milestone for the Linux Foundation mentorship on the Solang compiler.
For more information, please refer to the [project plan](https://wiki.hyperledger.org/display/INTERN/Project+plan+-+Solang+compiler+passes+2021).

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -140,6 +140,8 @@ pub fn parse_and_resolve(filename: &str, cache: &mut FileCache, target: Target)
         }
         Ok(file) => {
             sema::sema(file, cache, &mut ns);
+            sema::unused_variable::check_unused_namespace_variables(&mut ns);
+            sema::unused_variable::check_unused_events(&mut ns);
         }
```
Comment (seanyoung) [inline src/lib.rs:145]: It is better if `sema::sema()` calls the unused variable checks. I realise that `sema::sema()` is recursive, and you don't want this to be called for every recursive call. So maybe rename `sema::sema_file()` or something similar and create a new function `sema::sema()` which calls the unused variable checks.

File: src/sema/unused_variable.rs [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+        level: Level::Warning,
+        ty: ErrorType::Warning,
+        pos: Some(loc),
+        message: text.clone(),
+        notes
+    })
+}
+
+/// Emit different warning types according to the state variable usage
+pub fn emit_warning_local_variable(variable: &symtable::Variable) -> Option<Box<Diagnostic>> {
```
Comment (seanyoung) [inline src/sema/unused_variable.rs:178]: Why is the diagnositic in a box?

File: src/sema/unused_variable.rs [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+
+/// Marks as used variables that appear in an expression with right and left hand side.
+pub fn check_var_usage_expression(ns: &mut Namespace, left: &Expression, right: &Expression,
+                                  symtable: &mut Symtable) {
+    used_variable(ns, left, symtable);
+    used_variable(ns, right, symtable);
+}
+
+/// Generate warnings for unused varibles
+fn generate_unused_warning(loc: Loc, text: &String, notes: Vec<Note>) -> Box<Diagnostic> {
```
Comment (seanyoung) [inline src/sema/unused_variable.rs:167]: Why in a box?

**Review by seanyoung — CHANGES_REQUESTED** [review]
Looks great! Just some minor comments, I think other than some details it looks god.

I would say it is good practise to write tests as are you coding, so you have all the edge cases in mind and don't forget about them when writing tests.

**Comment by seanyoung** [comment]
Also please squash all your work into a single commit once you are done - or a few commits if each commit makes logical sense.

File: src/sema/unused_variable.rs (hunk 2 above)
Comment (LucasSte) [reply src/sema/unused_variable.rs:178]: I create the Diagnostic struct in function `generate_unused_warning` and it is returned to `emit_warning_local_variable`. There I return the value again to the original caller. I wanted to avoid memory copy between these three returns. I thought that using a smart pointer would be good.

File: src/lib.rs (hunk 1 above)
Comment (LucasSte) [reply src/lib.rs]: Fixed

File: src/sema/unused_variable.rs (hunk 2 above)
Comment (seanyoung) [reply src/sema/unused_variable.rs:214]: The only caller of `emit_warning_local_variable()` derefs the box with `*`, which effectively means unboxing it. Also using box for efficiency isn't the right thing to do: rust does a lot of inlining and tries to solve these issues for you.

Comment (LucasSte) [reply src/sema/unused_variable.rs]: Fixed

File: src/sema/unused_variable.rs [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+        | Expression::StorageArrayLength {
+            loc: _,
+            ty: _,
+            array,
+            ..
+        } => {
+            used_variable(ns, array, symtable);
+        }
+
+        Expression::StorageLoad(_, _, expr)
```
Comment (seanyoung) [inline src/sema/unused_variable.rs:91]: This function only recurses for limited set of expressions. For example, a simple `Expression::Add` is not mentioned. If you want to proper recursion there is a `recurse()` method on `Expression`.

However, as I read the code, I don't think this function needs to be recurse at all. sema resursively resolves expressions already.

File: src/sema/unused_variable.rs [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+
+/// Emit different warning types according to the state variable usage
+pub fn emit_warning_local_variable(variable: &symtable::Variable) -> Option<Diagnostic> {
+    match &variable.usage_type {
+        VariableUsage::Parameter => {
+            if !variable.read {
+                return Some(generate_unused_warning(
+                    variable.id.loc,
+                    &format!(
+                        "Function parameter '{}' has never been read.",
```
Comment (seanyoung) [inline src/sema/unused_variable.rs:234]: compiler diagnostics do not need a full stop or start with a capital letter

File: src/sema/unused_variable.rs [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+            }
+            None
+        }
+
+        VariableUsage::ReturnVariable => {
+            if !variable.assigned {
+                return Some(generate_unused_warning(
+                    variable.id.loc,
+                    &format!(
+                        "Return variable '{}' has never been assigned",
```
Comment (seanyoung) [inline src/sema/unused_variable.rs:248]: Same here and the others

File: src/sema/mod.rs [hunk 7]
```diff
@@ -39,9 +41,16 @@ pub const SOLANA_BUCKET_SIZE: u64 = 251;
...
+pub fn sema(file: ResolvedFile, cache: &mut FileCache, ns: &mut ast::Namespace) {
+    perform_sema(file, cache, ns);
+    check_unused_namespace_variables(ns);
+    check_unused_events(ns);
+}
+
 /// Load a file file from the cache, parse and resolve it. The file must be present in
 /// the cache. This function is recursive for imports.
-pub fn sema(file: ResolvedFile, cache: &mut FileCache, ns: &mut ast::Namespace) {
+pub fn perform_sema(file: ResolvedFile, cache: &mut FileCache, ns: &mut ast::Namespace) {
```
Comment (seanyoung) [inline src/sema/mod.rs:53]: `sema_file()` might be a better name.

File: src/sema/statements.rs [hunk 8]
```diff
@@ -309,12 +333,20 @@ fn statement(
...
             } else {
                 None
             };
 
-            if let Some(pos) = symtable.add(&decl.name, var_ty.clone(), ns) {
+            if let Some(pos) = symtable.add(
+                &decl.name,
+                var_ty.clone(),
+                ns,
+                initialized,
```
Comment (seanyoung) [inline src/sema/statements.rs:347]: You could replace this with `initializer.is_some()`

**Review by seanyoung — CHANGES_REQUESTED** [review]
Fantastic work!  I have some very minor comments.

Please also note that the clippy warnings need to be fixed. You can probably enable clippy in your ide.

File: .gitignore [hunk 9]
```diff
@@ -2,3 +2,5 @@
 Cargo.lock
 /target
 **/*.rs.bk
+.DS_Store
+.idea
```
[the remark this answers is not in the document]
Comment (LucasSte) [reply .gitignore:6]: `.idea` contains configuration files from my IDE. `.DS_Store` is a Mac OS file that holds meta-data about the folder it resides in. They are not supposed to be committed so I added them to gitignore. I'll remove them from this file and add to my global .gitignore then.

File: src/sema/unused_variable.rs (hunk 4 above)
Comment (LucasSte) [reply src/sema/unused_variable.rs:91]: Sema does resolve expressions recursively, as you mentioned. The problem is that using only `Expression::Variable` (and its variants) does not provide me the context in which the variable has been used, i.e., I cannot differentiate between an assignment and a read. Once I target that, I'll need to recurse other expressions that encapsulate a variable, such as `StorageLoad` and `ZeroExt` (but not limited to these two cases).

For `Expression::Add` and other expressions that have a left and a right side, I use the function `check_var_usage_expression`.

I agree that this function is worth some refactoring. I am happy to discuss more during our meeting tomorrow.

File: src/sema/unused_variable.rs [hunk 10]
```diff
@@ -0,0 +1,409 @@
...
+                        variable.id.name
+                    ),
+                    vec![],
+                ));
+            }
+            None
+        }
+
+        VariableUsage::StateVariable => {
```
Comment (LucasSte) [inline src/sema/unused_variable.rs:257]: State variables are not local variables. The name is misleading. Please, fix this @LucasSte.

Comment (LucasSte) [reply src/sema/unused_variable.rs]: Fixed

File: src/sema/statements.rs (hunk 8 above)
Comment (LucasSte) [reply src/sema/statements.rs]: Done

File: src/sema/mod.rs (hunk 7 above)
Comment (LucasSte) [reply src/sema/mod.rs]: Renamed!

File: src/sema/statements.rs [hunk 11]
```diff
@@ -685,6 +718,11 @@ fn statement(
                 return Err(());
             }
 
+            // for offset in symtable.returns.iter() {
```
[the remark this answers is not in the document]
Comment (LucasSte) [reply src/sema/statements.rs]: Done

File: src/sema/unused_variable.rs (hunk 4 above)
Comment (LucasSte) [reply src/sema/unused_variable.rs:86]: I simplified the function. Please, take a look again. I tried to recurse only for expressions that directly wrap a variable.

File: src/sema/mod.rs [hunk 12]
```diff
@@ -39,9 +41,16 @@ pub const SOLANA_BUCKET_SIZE: u64 = 251;
...
+pub fn sema(file: ResolvedFile, cache: &mut FileCache, ns: &mut ast::Namespace) {
+    sema_file(file, cache, ns);
+    check_unused_namespace_variables(ns);
+    check_unused_events(ns);
+}
+
 /// Load a file file from the cache, parse and resolve it. The file must be present in
 /// the cache. This function is recursive for imports.
-pub fn sema(file: ResolvedFile, cache: &mut FileCache, ns: &mut ast::Namespace) {
+pub fn sema_file(file: ResolvedFile, cache: &mut FileCache, ns: &mut ast::Namespace) {
```
Comment (seanyoung) [inline src/sema/mod.rs:53]: sema_fiile should no longer be `pub`

File: src/sema/mod.rs [hunk 13]
```diff
@@ -39,9 +41,16 @@ pub const SOLANA_BUCKET_SIZE: u64 = 251;
 pub const SOLANA_FIRST_OFFSET: u64 = 16;
 pub const SOLANA_SPARSE_ARRAY_SIZE: u64 = 1024;
 
+/// Performs semantic analysis and checks for unused variables
```
Comment (seanyoung) [inline src/sema/mod.rs:44]: The comment is not really correct. This function is the entry point for sema. Maybe move the comment from sema_file to here and put this comment inside the function.

File: src/sema/statements.rs [hunk 14]
```diff
@@ -1236,7 +1295,9 @@ fn destructure_values(
                 diagnostics,
                 Some(&Type::Bool),
             )?;
-
+            std::println!("Aqui!!");
+            std::println!("{:?}", cond);
```
Comment (seanyoung) [inline src/sema/statements.rs:1299]: Please remove your debugging :smiley:

File: src/sema/variables.rs [hunk 15]
```diff
@@ -250,6 +252,8 @@ pub fn var_decl(
         ty: ty.clone(),
         constant: is_constant,
         initializer,
+        read: false,
+        assigned: initialized,
```
Comment (seanyoung) [inline src/sema/variables.rs:256]: `initializer.is_some()` ?

**Review by seanyoung — CHANGES_REQUESTED** [review]
Looks good, just a few very minor comments.

File: src/sema/mod.rs (hunk 12 above)
Comment (LucasSte) [reply src/sema/mod.rs]: Done

File: src/sema/mod.rs (hunk 13 above)
Comment (LucasSte) [reply src/sema/mod.rs]: I moved the comments and added a new one to `sema_file`

File: src/sema/statements.rs (hunk 14 above)
Comment (LucasSte) [reply src/sema/statements.rs]: As a curiosity, `aqui` means `here` in Portuguese :)

File: src/sema/variables.rs (hunk 15 above)
Comment (LucasSte) [reply src/sema/variables.rs]: I changed `initializer` to `initializer: initializer.clone()` as well.

**Comment by seanyoung** [comment]
Would you mind rebasing against latest main, so it can be merged and also squashing the commits into one. Thanks.

**Comment by LucasSte** [comment]
> Would you mind rebasing against latest main, so it can be merged and also squashing the commits into one. Thanks.

I left two commits: one for the code itself and another for the tests.

**Review by seanyoung — APPROVED** [review]
